### DISCUSSION TITLE
Access Modifier: "namespace internal"

### DISCUSSION THREAD

**Discussion by Jaelid** [dotnet/csharplang#8224] [Language Ideas]
On occasion I find I want to suggest to the user they shouldn't use a class as its designed for use in a specific scope. This is great if you are writing a library, just mark it as a internal and now when people reference that library its not accessible. However this is not so easily done if you're writing the library itself or an application. Instead you can nest the class inside another and make it private however now if there are several classes the file becomes big and you hit a wall if one of those classes needs to be accessed by just one other class in the same scope. So the suggestion is

internal namespace: access is limited to the same namespace

**potentially**
I considered "protected internal namespace" but wasn't sure what the would fully entail if talking about derived classes. Perhaps someone smarter than me can consider this

**Comment by Richiban** [comment]
This has been requested several times before... One of the principal issues is that namespaces don't really exist in dotnet; the namespace is just a prefix to the type name. This makes it trivial to simply move your type to a different namespace to gain access to a type that you shouldn't have.

The answer here is an analyzer; I'm sure one exists already that will give the ability to define namespace-level accessibility or friend classes.

**Reply by Jaelid** [reply]
Yes, that makes sense. I do quite like the idea of moving something to a namespace to gain access to the type. It seems like a conscious decision, similar to inheriting from a class with internal types and exposing access to them is intentional essentially still making extensible but not modifiable. I am glad I'm not the first to request this though so thanks for pointing that out. Smart suggestion on the analyser I will go off and do some reading as I have never written one, hoping they give edit time feedback 👍

**Reply by hughesjs** [reply]
> This has been requested several times before... One of the principal issues is that namespaces don't really exist in dotnet; the namespace is just a prefix to the type name. This makes it trivial to simply move your type to a different namespace to gain access to a type that you shouldn't have.
> 
> The answer here is an analyzer; I'm sure one exists already that will give the ability to define namespace-level accessibility or friend classes.

 You can use reflection to get access to anything you want (at least to my understanding). They're really there to enforce coding practice more than anything rather than stop you accessing something you shouldn't. So I don't see this being a reason why this feature shouldn't be added.

**Reply by KennethHoff** [reply]
> This has been requested several times before... One of the principal issues is that namespaces don't really exist in dotnet; the namespace is just a prefix to the type name.

The same exact argument can be used for the accessibility modifier `file` - files are not a concept in .Net - yet they were added.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/8042 https://github.com/dotnet/csharplang/issues/6794
